inference-engine/src/mkldnn_plugin/nodes/mkldnn_pad_node.cpp

```diff
@@ -95,158 +203,312 @@ void MKLDNNPadNode::createPrimitive() {
+        if (padsBegin[i] != 0 || padsEnd[i] != 0) {
+            endIdx = i;
+            break;
+        }
+    }
 
+    size_t nGluingLastDims = params.dstStrides[std::max(endIdx - 1, 0)];
+    params.nDimsForWork = std::max(endIdx - std::max(beginIdx, 0), 1);
+    params.workAmount = params.dstDims[0];
+    for (int i = 1; i <= beginIdx; ++i) {
+        params.workAmount *= params.dstDims[i];
+        params.dstDims[0] *= params.dstDims[i];
+        params.srcDims[0] *= params.srcDims[i];
+        params.dstStrides[0] /= params.dstDims[i];
+        params.srcStrides[0] /= params.srcDims[i];
+    }
+
+    if (beginIdx > 0) {
+        beginIdx++;
+        params.dstDims.erase(params.dstDims.begin() + 1, params.dstDims.begin() + beginIdx);
+        params.srcDims.erase(params.srcDims.begin() + 1, params.srcDims.begin() + beginIdx);
+        params.dstStrides.erase(params.dstStrides.begin() + 1, params.dstStrides.begin() + beginIdx);
+        params.srcStrides.erase(params.srcStrides.begin() + 1, params.srcStrides.begin() + beginIdx);
+        padsBegin.erase(padsBegin.begin() + 1, padsBegin.begin() + beginIdx);
+        padsEnd.erase(padsEnd.begin() + 1, padsEnd.begin() + beginIdx);
+    }
+    params.workAmount = params.workAmount * params.dstStrides[0] / nGluingLastDims;
+
+    params.lastDstDim = nGluingLastDims;
+    params.shift = params.dstStrides[params.nDimsForWork];
+    if (padMode != CONSTANT || (padMode == CONSTANT && padValue == 0)) {
+        params.lastDstDim *= params.sizeData;
+        params.shift *= params.sizeData;
+    }
+
+    for (size_t i = 0; i < params.srcDims.size(); ++i)
+        params.srcODims.push_back(padsBegin[i] + params.srcDims[i]);
+
+    if (padMode == REFLECT || padMode == SYMMETRIC) {
+        int shift = padMode == SYMMETRIC ? 1 : 0;
+        for (size_t i = 0; i < params.srcDims.size(); ++i)
+            params.srcDimsForReflectOrSymmetric.push_back(params.srcDims[i] + params.srcODims[i] - 2 + shift);
+    }
+}
+
+void MKLDNNPadNode::execute(mkldnn::stream strm) {
     switch (padMode) {
         case CONSTANT:
-            padConstantOrEdge(srcData, dstData, srcDims, dstDims);
+            padConstant();
             break;
         case EDGE:
-            padConstantOrEdge(srcData, dstData, srcDims, dstDims, true);
+            padEdge();
             break;
         case REFLECT:
-            padReflectOrSymmetric(srcData, dstData, srcDims, dstDims);
+            padReflectOrSymmetric();
             break;
         case SYMMETRIC:
-            padReflectOrSymmetric(srcData, dstData, srcDims, dstDims, true);
+            padReflectOrSymmetric(true);
             break;
     }
 }
 
-inline size_t parallel_init(size_t start, size_t size, std::vector<size_t> &counters, std::vector<size_t> &dims) {
-    for (int j = size - 1; j >= 0; j--) {
-        counters[j] = start % dims[j];
+inline size_t parallel_init(size_t start, size_t nDims, const SizeVector& dims, SizeVector& indexes) {
```

Comment:
Please specify such functions as `static` to restrict their scope.

Reply:
Thanks, done